File: 20231018_Wednesday/20231018.rb
# Leetcode: 2050. Parallel Courses III.
# https://leetcode.com/problems/parallel-courses-iii/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 280 ms, faster than 100.00% of Ruby online submissions for Parallel
# Courses III.
# Memory Usage: 247.6 MB, less than 100.00% of Ruby online submissions for
# Parallel Courses III.
# 2023.10.18 Daily Challenge.
# @param {Integer} n
# @param {Integer[][]} relations
# @param {Integer[]} time
# @return {Integer}
def minimum_time(n, relations, time)
  # Based on:
  # https://leetcode.com/problems/parallel-courses-iii/solution/
  @g = {}
  @dp = {}
  @t = time
  n.times do |i|
    @g[i] = []
  end
  relations.each do |e|
    x,y = e[0] - 1, e[1] - 1
    @g[x].push(y)
  end
  ans = 0
  n.times do |n|
    tmp = dfs(n)
    ans = tmp if tmp > ans
  end
  return ans
end

def dfs(n)
  if @dp[n]
    return @dp[n]
  end
  if ! @g[n] 
    return @t[n]
  end
  if 0 == @g[n].size
    return @t[n]
  end
  ans = 0
  @g[n].each do |nb|
    tmp = dfs(nb)
    ans = tmp if tmp > ans
  end
  @dp[n] = @t[n] + ans
  return @t[n] + ans
end

File: 20231018_Wednesday/20231018.go
// Leetcode: 2050. Parallel Courses III.
// https://leetcode.com/problems/parallel-courses-iii/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 359 ms, faster than 50.00% of Go online submissions for Parallel
// Courses III.
// Memory Usage: 22.6 MB, less than 50.00% of Go online submissions for Parallel
// Courses III.
// 2023.10.18 Daily Challenge.
package main

var g map[int][]int
var dp map[int]int
var t []int

func minimumTime(n int, relations [][]int, time []int) int {
	// Based on:
	// https://leetcode.com/problems/parallel-courses-iii/solution/
	g = make(map[int][]int)
	dp = make(map[int]int)
	t = time
	for i := 0; i < n; i += 1 {
		g[i] = make([]int, 0)
	}
	for _, e := range relations {
		x, y := e[0]-1, e[1]-1
		g[x] = append(g[x], y)
	}
	ans := 0
	for nd := 0; nd < n; nd += 1 {
		tmp := dfs(nd)
		if ans < tmp {
			ans = tmp
		}
	}
	return ans
}

func dfs(n int) int {
	if _, ok := dp[n]; ok {
		return dp[n]
	}
	if _, ok := g[n]; !ok {
		return t[n]
	}
	if 0 == len(g[n]) {
		return t[n]
	}
	ans := 0
	for _, nb := range g[n] {
		tmp := dfs(nb)
		if tmp > ans {
			ans = tmp
		}
	}
	ans += t[n]
	dp[n] = ans
	return ans
}
